### PR TITLE
Android lifecycle state tracking and message handling

### PR DESCRIPTION
**Pull request by GaryQian**
Android-side support for fix for https://github.com/flutter/flutter/issues/2716

This adds a reply to LifecycleChannel which allows Dart to request Android to send the current lifecycle state through the channel.

This was implemented inside LifecycleChannel as it consolidates all lifecycle updates to one channel. The message passed is ignored, and will both reply with the state as well as send through the existing lifecycle state message channel.

In order to achieve the desired behavior, the query should be made right in the `runApp` call in the framework.

iOS version is under https://github.com/flutter/engine/pull/7931

### CODE REVIEW

**Comment by matthew-carroll** [comment]
My initial thought is that we shouldn't be querying this across the channel boundary.  Why isn't Flutter just saving the last state that was communicated to it and then making that available?

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java [hunk 1]
```diff
@@ -15,23 +16,69 @@
...
 
+  public enum AppLifecycleState { INACTIVE, RESUMED, PAUSED }
+
   @NonNull
   public final BasicMessageChannel<String> channel;
 
+  /**
+   * Cache the current state so that any queries for it may access it.
+   */
+  private AppLifecycleState state = AppLifecycleState.INACTIVE;
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java]: Please keep all system channels stateless.  The purpose of these channels is to codify the comm API, not add behavior.  Please follow examples of `TextInputPlugin`, `AndroidKeyProcessor`, and other areas where state surrounds a channel.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java [hunk 2]
```diff
@@ -5,6 +5,7 @@
 package io.flutter.embedding.engine.systemchannels;
 
 import android.support.annotation.NonNull;
+import android.util.*;
```
Comment (GaryQian) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java]: Will change to actual imports.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java [hunk 3]
```diff
@@ -15,23 +16,69 @@
  */
 public class LifecycleChannel {
 
+  public enum AppLifecycleState { INACTIVE, RESUMED, PAUSED }
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java]: Would you mind moving all enums and inner classes to the bottom of the file?

Also, please add comments for each enum value that describe what exactly they represent in terms of lifecycle.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java [hunk 4]
```diff
@@ -15,23 +16,69 @@
...
+    sendCurrentState();
   }
 
   public void appIsPaused() {
-    channel.send("AppLifecycleState.paused");
+    state = AppLifecycleState.PAUSED;
+    sendCurrentState();
   }
 
+  public static String lifecycleStateToString(AppLifecycleState state) {
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java]: Please add this to the enum definition, similar to the way all other channels with enums serialize and deserialize.

**Comment by GaryQian** [comment]
This PR is on hold, will likely be superseded by an engine-level windowdata cache based system.

**Pull request closed without merging**